### PR TITLE
Base and Weather Ingress

### PR DESCRIPTION
**Pull request by EdwinB12**
Includes: 

- BaseIngress for shared utilities. Class to be inherited by specific data ingress options. 
- Readme describing what the BaseIngress is and how to use it to write custom DataIngress Classes. 
- OpenWeatherDataIngress Class for ingressing weather data from OpenWeatherMap API. 
- Basic tests for handling dt_from and dt_to. These are the variables determining what time range to pul data from. 

TODO before merging : 

- Tests for BaseClass methods. Will require a simple mocking of the `get_data` method.  
- Tests for OpenWeatherDataIngress methods 

TODO but perhaps different PR: 

- Azure Functions Utility. Hopefully can be included into BaseIngress

### CODE REVIEW

File: dtbase/ingress/README.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+- *SENSOR_TYPE:* Define the sensor type as detailed by the "/sensor/insert-sensor-type" API endpoint.
+- *SENSOR_OPENWEATHERMAPHISTORICAL*: Define the sensor as detailed by the "/sensor/insert-sensor" API endpoint. This sensor is for the historical data API.
+- *SENSOR_OPENWEATHERMAPFORECAST*: Same as previous sensor but for forecasting.
+
+These constants are dictionaries and define the sensor type and sensor payloads. The payload for entering sensor-readings is built in the get_data method.
+
+### 2. OpenWeatherDataIngress
+
+We then write a custom class that inherits from BaseIngress. There are a number of _* methods that are used to handle different combinations of start and end dates given by the user. A lot of this complexity comes from their being two different APIs.
```
Comment (mhauru) [inline dtbase/ingress/README.md:86]:
```suggestion
We then write a custom class that inherits from BaseIngress. There are a number of _* methods that are used to handle different combinations of start and end dates given by the user. A lot of this complexity comes from there being two different APIs.
```

**Comment by mhauru** [comment]
>TODO but perhaps different PR:
    Azure Functions Utility. Hopefully can be included into BaseIngress

Yeah, agreed, this can be a separate PR.

**Comment by EdwinB12** [comment]
@mhauru - I'm reading in some hardcoded responses for mocking purposes (such as MOCKED_CONST_OPENWEATHERMAP_HISTORICAL_URL_RESPONSE). They are quite large dictionaries. Have you got any strong opinions on where I should store these dictionaries: 

- In test_ingress_weather.py 
- In conftest.py (Have them as fixtures). 
- In a seperate resources file (this is what I've currently implemented).

**Comment by mhauru** [comment]
The separate resource files sound good to me.

**Review by mhauru — CHANGES_REQUESTED** [review]
Looks great. I had a number of small suggestions, I put them in a PR that is asking to be merged to your branch: https://github.com/alan-turing-institute/DTBase/pull/178

**Comment by mhauru** [comment]
One thing I was thinking of was that maybe there should be a separate pair of OpenWeatherMap sensors (historical and forecast) for every lat/long. That the lat and the long would be in the unique identifier of the sensor. Maybe this is for a separate PR though. I'm in general not fully happy with how we currently use environment variables, there may be a bit of an overhaul to do there, but it might make sense to do once things are running on Azure, since envvars are often the best way to handle parameters there.

**Review by mhauru — APPROVED** [review]

**Pull request merged**